websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,581 @@
+    use crate::golem::web_search::web_search::{Guest, SearchSession};
+    use crate::golem::web_search::web_search::{
+        SearchError, SearchMetadata, SearchParams, SearchResult,
+    };
+
+    impl<Impl: ExtendedwebsearchGuest> Guest for Durablewebsearch<Impl> {
+        type SearchSession = Impl::SearchSession;
+
+        fn start_search(params: SearchParams) -> Result<SearchSession, SearchError> {
+            Impl::start_search(params)
+        }
+
+        fn search_once(
+            params: SearchParams,
+        ) -> Result<(Vec<SearchResult>, Option<SearchMetadata>), SearchError> {
+            Impl::search_once(params)
+        }
+    }
+}
+
+/// When the durability feature flag is on, wrapping with `Durablewebsearch` adds custom durability
+/// on top of the provider-specific websearch implementation using Golem's special host functions and
+/// the `golem-rust` helper library.
+///
+/// There will be custom durability entries saved in the oplog, with the full websearch request and configuration
+/// stored as input, and the full response stored as output. To serialize these in a way it is
+/// observable by oplog consumers, each relevant data type has to be converted to/from `ValueAndType`
+/// which is implemented using the type classes and builder in the `golem-rust` library.
+#[cfg(feature = "durability")]
+mod durable_impl {
+    use crate::durability::{Durablewebsearch, ExtendedwebsearchGuest};
+    use crate::exports::golem::web_search::web_search::{Guest, GuestSearchSession, SearchSession};
+    use crate::exports::golem::web_search::web_search::{
+        SearchError, SearchMetadata, SearchParams, SearchResult,
+    };
+    use golem_rust::bindings::golem::durability::durability::{
+        DurableFunctionType, LazyInitializedPollable,
+    };
+    use golem_rust::durability::Durability;
+    use golem_rust::wasm_rpc::Pollable;
+    use golem_rust::{with_persistence_level, FromValueAndType, IntoValue, PersistenceLevel};
+    use std::cell::RefCell;
+    use std::fmt::{Display, Formatter};
+
+    impl<Impl: ExtendedwebsearchGuest> Guest for Durablewebsearch<Impl> {
+        type SearchSession = DurableSearchSession<Impl>;
+
+        fn start_search(params: SearchParams) -> Result<SearchSession, SearchError> {
+            let durability = Durability::<SearchParams, UnusedError>::new(
+                "golem_websearch",
+                "start_search",
+                DurableFunctionType::WriteRemote,
+            );
+
+            if durability.is_live() {
+                let result = with_persistence_level(PersistenceLevel::PersistNothing, || {
+                    match Impl::start_search(params.clone()) {
+                        Ok(_session) => Ok(params.clone()),
+                        Err(e) => Err(e),
+                    }
+                });
+
+                let persisted_params = result?;
+                durability.persist_infallible(
```

Comment:
As the errors are part of the api, they should be persisted as part of replays (i.e. the approach that golem takes is that it does automatic replay on unexpected errors, expected errors should be persisted and threated the same as a successful result).

So please change all of these to use persist instead of persist_infalliable